vm_args.c

```diff
@@ -779,25 +779,41 @@ argument_kw_error(rb_execution_context_t *ec, const rb_iseq_t *iseq, const char
 }
 
 static inline void
-vm_caller_setup_arg_splat(rb_control_frame_t *cfp, struct rb_calling_info *calling)
+vm_caller_setup_arg_splat(rb_control_frame_t *cfp, struct rb_calling_info *calling, int allow_heap_argv)
 {
     int argc = calling->argc;
     VALUE *argv = cfp->sp - argc;
     VALUE ary = argv[argc-1];
 
     vm_check_canary(GET_EC(), cfp->sp);
-    cfp->sp--;
 
     if (!NIL_P(ary)) {
         const VALUE *ptr = RARRAY_CONST_PTR_TRANSIENT(ary);
-        long len = RARRAY_LEN(ary), i;
-
-        CHECK_VM_STACK_OVERFLOW(cfp, len);
+        long len = RARRAY_LEN(ary);
+
+        if (allow_heap_argv && len + argc > VM_ARGC_STACK_MAX) {
+            /* Avoid SystemStackError when calling cfuncs with splatted large arrays by storing
+             * argv in a temporary array, instead of trying to keeping it on the VM stack.
+             */
+            VALUE argv_array = rb_ary_new_capa(len + argc - 1);
```

Comment:
JFYI you might want to use rb_alloc_tmp_buffer().
https://github.com/ruby/ruby/blob/ee49fe5d34c1498a532bbaeaf4c12b162d0df78f/gc.c#L12600